File: thirdmaximumnumber.java
// Return the third distinct maximum number in array, if does not exist, return the maximum number.

class Solution {
    public int thirdMax(int[] nums) {
        Integer max = null;
        Integer secondMax = null;
        Integer thirdMax = null;

        for (Integer num : nums) {
            if (num.equals(max) || num.equals(secondMax) || num.equals(thirdMax)) {
                continue;
            }
            if (max == null || num > max) {
                thirdMax = secondMax;
                secondMax = max;
                max = num;
            } else if (secondMax == null || num > secondMax) {
                thirdMax = secondMax;
                secondMax = num;
            } else if (thirdMax == null || num > thirdMax) {
                thirdMax = num;
            }
        }
        return thirdMax != null ? thirdMax : max;
    }
}

File: thirdmaximumnumber.cpp
// Return the third distinct maximum number in array, if does not exist, return the maximum number.

#include <vector>
#include <set>
using namespace std;

class Solution {
public:
    int thirdMax(vector<int>& nums) {
        set<int> top3;

        for (int num : nums) {
            top3.insert(num);

            if (top3.size() > 3) {
                top3.erase(top3.begin());
            }
        }

        return top3.size() == 3 ? *top3.begin() : *top3.rbegin();
    }
};
